### PR TITLE
[MM-39629] - Migrate from gorp to sqlx in store/sqlstore/thread_store.go

### PR DESCRIPTION
**Pull request by penthaapatel**
#### Summary

Migrate from `gorp` to `sqlx` in `store/sqlstore/thread_store.go`
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/18867
JIRA https://mattermost.atlassian.net/browse/MM-39629
#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Hey @penthaapatel - how is it going with this? Let me know if I can help.

**Comment by penthaapatel** [comment]
Hi @agnivade! Sorry this one is taking a little bit longer than usual as I got an arm injury, have been away from keyboard since then. Just one error to fix I think, will do it ASAP.

**Comment by agnivade** [comment]
Woops, sorry to hear that! No rush on this, please rest up and work only when you feel better!

**Comment by penthaapatel** [comment]
Thank you @agnivade! :)

**Comment by isacikgoz** [comment]
Get well soon @penthaapatel 🙏

**Comment by penthaapatel** [comment]
Thank you @isacikgoz! :)

**Comment by penthaapatel** [comment]
Hi @agnivade! I need some help, could you please suggest me the possible solutions to fix the failing tests?

1. 68e366ea1575456e7085aba55c81c178144cd1a3 - I tried fixing the syntax error but it still is incorrect.
Error generated - `pq: invalid input syntax for type json`
2. There is a Scan error when reading data from the `ThreadMemberships` table in the database, not quite sure how to fix this one.

**Comment by agnivade** [comment]
Good to see you back @penthaapatel ! And sorry for the delay on this one.

1. Indeed, this is a tricky problem. I see what you have tried to do though. You have replaced all `$`s by `?`s. But then the query itself needs `?` to work with. That's just how json operator syntax works with Postgres.

Let's try to understand the situation a bit better. We need a `?` to be there in the query, but it's not a parameter, rather a part of the query itself. Which means we need sqlx to escape it. But unfortunately, it doesn't:

> // FIXME: this should be able to be tolerant of escaped ?'s in queries without
// losing much speed, and should be to avoid confusion.

https://github.com/jmoiron/sqlx/blob/a62bc6088664eb854c3d5ebbd08a1ffd59662ef1/bind.go#L56

So then what are our options? You need to understand that any `?` for postgres queries eventually get converted to `$1`, `$2` again, because that's how postgres accepts parameters. But then it does it for _all_ queries. So what we need to tell sqlx is to stop doing it for this query. Therefore, a simple solution is to add another method which avoids this rebind business.

If you can add this diff to your PR, it should fix the problem:

```diff
diff --git a/store/sqlstore/sqlx_wrapper.go b/store/sqlstore/sqlx_wrapper.go
index 08e998a07..edbcdcc6d 100644
--- a/store/sqlstore/sqlx_wrapper.go
+++ b/store/sqlstore/sqlx_wrapper.go
@@ -102,6 +102,12 @@ func (w *sqlxDBWrapper) NamedExec(query string, arg interface{}) (sql.Result, er
 func (w *sqlxDBWrapper) Exec(query string, args ...interface{}) (sql.Result, error) {
        query = w.DB.Rebind(query)

+       return w.ExecRaw(query, args...)
+}
+
+// ExecRaw is like Exec but without any rebinding of params. You need to pass
+// the exact param types of your target database.
+func (w *sqlxDBWrapper) ExecRaw(query string, args ...interface{}) (sql.Result, error) {
        ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
        defer cancel()

@@ -203,6 +209,12 @@ func (w *sqlxTxWrapper) Get(dest interface{}, query string, args ...interface{})
 func (w *sqlxTxWrapper) Exec(query string, args ...interface{}) (sql.Result, error) {
        query = w.Tx.Rebind(query)

+       return w.ExecRaw(query, args...)
+}
+
+// ExecRaw is like Exec but without any rebinding of params. You need to pass
+// the exact param types of your target database.
+func (w *sqlxTxWrapper) ExecRaw(query string, args ...interface{}) (sql.Result, error) {
        ctx, cancel := context.WithTimeout(context.Background(), w.queryTimeout)
        defer cancel()

diff --git a/store/sqlstore/thread_store.go b/store/sqlstore/thread_store.go
index b3ab328a3..5622abdc3 100644
--- a/store/sqlstore/thread_store.go
+++ b/store/sqlstore/thread_store.go
@@ -746,10 +746,10 @@ func (s *SqlThreadStore) MaintainMembership(userId, postId string, opts store.Th

        if opts.UpdateParticipants {
                if s.DriverName() == model.DatabaseDriverPostgres {
-                       if _, err2 := trx.Exec(`UPDATE Threads
-                               SET participants = participants || ?::jsonb
-                               WHERE postid=?
-                               AND NOT participants=?`, jsonArray([]string{userId}), postId, userId); err2 != nil {
+                       if _, err2 := trx.ExecRaw(`UPDATE Threads
+                               SET participants = participants || $1::jsonb
+                               WHERE postid=$2
+                               AND NOT participants ? $3`, jsonArray([]string{userId}), postId, userId); err2 != nil {
                                return nil, err2
                        }
                } else {
```

Hopefully, this was clear. Feel free to ping me on Community if you don't understand any part of it.

2. Can you point me to where you are seeing the error?

**Comment by penthaapatel** [comment]
Hi @agnivade! :)

1. Yes, I understand the issue now. Thank you so much for the clear explanation and helping me out! :) I will soon push the changes to fix it.
2. The following errors occur, `sql: Scan error` is common to all of these errors. 
```
{"timestamp":"2022-01-20 19:01:10.032 Z","level":"error","msg":"Unable to update the last viewed at time.","caller":"web/context.go:105","path":"/api/v4/channels/members/cimo76s9nffzfx3qfm6gmz8gay/view","request_id":"ejpioeahjfna3jxfb1wzaxukwa","ip_addr":"127.0.0.1","user_id":"cimo76s9nffzfx3qfm6gmz8gay","method":"POST","err_where":"MarkChannelsAsViewed","http_code":500,"err_details":"failed to get thread membership with userid=cimo76s9nffzfx3qfm6gmz8gay: sql: no rows in result set"}
{"timestamp":"2022-01-20 19:01:10.032 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"POST","url":"/api/v4/channels/members/cimo76s9nffzfx3qfm6gmz8gay/view","request_id":"ejpioeahjfna3jxfb1wzaxukwa","status_code":"500"}
    channel_test.go:2382: 
        	Error Trace:	channel_test.go:2382
        	Error:      	Received unexpected error:
        	            	: Unable to update the last viewed at time., 
        	Test:       	TestViewChannel
```
```
{"timestamp":"2022-01-20 19:01:44.561 Z","level":"error","msg":"Unable to update the last viewed at time.","caller":"web/context.go:105","path":"/api/v4/channels/members/cimo76s9nffzfx3qfm6gmz8gay/view","request_id":"78msjery77g3tnmx6g87b35sdh","ip_addr":"127.0.0.1","user_id":"cimo76s9nffzfx3qfm6gmz8gay","method":"POST","err_where":"MarkChannelsAsViewed","http_code":500,"err_details":"failed to get thread membership with userid=cimo76s9nffzfx3qfm6gmz8gay: sql: Scan error on column index 0, name \"PostId\": unsupported Scan, storing driver.Value type []uint8 into type *[]string"}
{"timestamp":"2022-01-20 19:01:44.561 Z","level":"debug","msg":"Received HTTP request","caller":"web/handlers.go:156","method":"POST","url":"/api/v4/channels/members/cimo76s9nffzfx3qfm6gmz8gay/view","request_id":"78msjery77g3tnmx6g87b35sdh","status_code":"500"}
    channel_test.go:4474: 
        	Error Trace:	channel_test.go:4474
        	Error:      	Received unexpected error:
        	            	: Unable to update the last viewed at time., 
        	Test:       	TestViewChannelWithoutCollapsedThreads
```
```
    post_test.go:2615: 
        	Error Trace:	post_test.go:2615
        	Error:      	Expected nil, but got: &model.AppError{Id:"app.channel.update_last_viewed_at.app_error", Message:"Unable to update the last viewed at time.", DetailedError:"failed to get thread membership with userid=jj19or3bztgidyytas8u6fencr: sql: no rows in result set", RequestId:"", StatusCode:500, Where:"MarkChannelsAsViewed", IsOAuth:false, params:map[string]interface {}(nil)}
        	Test:       	TestSetChannelUnread
```
```
    channel_test.go:2174: 
        	Error Trace:	channel_test.go:2174
        	Error:      	Expected nil, but got: &model.AppError{Id:"app.channel.update_last_viewed_at.app_error", Message:"Unable to update the last viewed at time.", DetailedError:"failed to get thread membership with userid=mbp445jukfn9dmf4mxs84bjcyy: sql: Scan error on column index 0, name \"PostId\": unsupported Scan, storing driver.Value type []uint8 into type *[]string", RequestId:"", StatusCode:500, Where:"MarkChannelsAsViewed", IsOAuth:false, params:map[string]interface {}(nil)}
        	Test:       	TestViewChannelCollapsedThreadsTurnedOff
```

**Comment by agnivade** [comment]
> The following errors occur, sql: Scan error is common to all of these errors.

I will look into that. Can you push the changes for #1 and also change all nil slice occurrences into empty slices. (change `var name []string` to `name := []string{}`)>

**Comment by agnivade** [comment]
Thanks @penthaapatel. There was only a slight mistake `.Get()` is meant to return only a single row. But the original query expected multiple rows, so we need to use `.Select()`. This should fix it

```diff
diff --git a/store/sqlstore/thread_store.go b/store/sqlstore/thread_store.go
index d1cc79629..34eebf62a 100644
--- a/store/sqlstore/thread_store.go
+++ b/store/sqlstore/thread_store.go
@@ -504,7 +504,7 @@ func (s *SqlThreadStore) MarkAllAsReadInChannels(userID string, channelIDs []str
                Where(sq.Eq{"ThreadMemberships.UserId": userID}).
                ToSql()

-       err := s.GetReplicaX().Get(&threadIDs, query, args...)
+       err := s.GetReplicaX().Select(&threadIDs, query, args...)
        if err != nil {
                return errors.Wrapf(err, "failed to get thread membership with userid=%s", userID)
        }
```

**Review by agnivade — APPROVED** [review]
Incredible work @penthaapatel !

**Comment by penthaapatel** [comment]
> Incredible work @penthaapatel !

Thank you @agnivade for all the help! :)

File: store/sqlstore/thread_store.go [hunk 1]
```diff
@@ -83,25 +85,39 @@ func (s *SqlThreadStore) Save(thread *model.Thread) (*model.Thread, error) {
...
 
 func (s *SqlThreadStore) Update(thread *model.Thread) (*model.Thread, error) {
-	return s.update(s.GetMaster(), thread)
-}
-
-func (s *SqlThreadStore) update(ex gorp.SqlExecutor, thread *model.Thread) (*model.Thread, error) {
-	if _, err := ex.Update(thread); err != nil {
+	jsonParticipants, err := json.Marshal(thread.Participants)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed marshalling thread participants")
```
Comment (isacikgoz) [inline store/sqlstore/thread_store.go:90]:
```suggestion
		return nil, errors.Wrap(err, "failed marshaling thread participants")
```

File: store/sqlstore/thread_store.go [hunk 2]
```diff
@@ -610,7 +664,17 @@ func (s *SqlThreadStore) getMembershipForUser(ex gorp.SqlExecutor, userId, postI
...
+		Where(sq.And{
+			sq.Eq{"PostId": postId},
+			sq.Eq{"UserId": userId},
+		}).
+		ToSql()
+	if err != nil {
+		return errors.Wrap(err, "threadmembership_tosql")
+	}
+	if _, err := s.GetMasterX().Exec(query, args...); err != nil {
 		return errors.Wrap(err, "failed to update thread membership")
```
Comment (isacikgoz) [inline store/sqlstore/thread_store.go:678]: Since we are here
```suggestion
		return errors.Wrap(err, "failed to delete thread membership")
```

**Review by isacikgoz — APPROVED** [review]
Looks good, thank you very much @penthaapatel 💐 I have just a non-blocking couple of small nits.

**Comment by penthaapatel** [comment]
Hi @isacikgoz! Just pushed the suggested changes, hope it's good to go now. :)

**Pull request merged**